Source/NSGridView.m

```diff
@@ -0,0 +1,1232 @@
+                  [gc setGridView: self];                  
+                }
+              else
+                {
+                  gc = [_columns objectAtIndex: c];
+                }
+
+              [cell _setOwningRow: gr];
+              [cell _setOwningColumn: gc];
+              [cell setContentView: v];
+              
+              [_cells addObject: cell];
+              c++;
+            }
+          END_FOR_IN(row);
+          r++;
+        }
+      END_FOR_IN(views);
+      [self _refreshCells];
+    }  
+  return self;
+}
+
++ (instancetype) gridViewWithNumberOfColumns: (NSInteger)columnCount rows: (NSInteger)rowCount
+{
+  NSUInteger r = 0;
+  NSUInteger c = 0;
+  NSMutableArray *rows = [[NSMutableArray alloc] initWithCapacity: rowCount];
+  
+  for (r = 0; r < rowCount; r++)
+    {
+      NSMutableArray *col = [[NSMutableArray alloc] initWithCapacity: columnCount];
+      for (c = 0; c < columnCount; c++)
+        {
+          NSView *gv = [[NSView alloc] init];
+          [col addObject: gv];
+          RELEASE(gv);
+        }
+      [rows addObject: col];
+    }
+
+  return [self gridViewWithViews: rows];
+}
+
++ (instancetype) gridViewWithViews: (NSArray *)views
+{
+  return AUTORELEASE([[self alloc] initWithViews: views]);
+}
+
+- (void) setFrame: (NSRect)f
+{
+  [super setFrame: f];
+  [self _refreshCells];
+}
+
+- (NSInteger) numberOfRows
+{
+  return [_rows count];
+}
+
+- (NSInteger) numberOfColumns
+{
+  return [_columns count];
+}
+
+- (NSGridRow *) rowAtIndex: (NSInteger)index
+{
+  return [_rows objectAtIndex: index];
+}
+
+- (NSInteger) indexOfRow: (NSGridRow *)row
+{
+  return [_rows indexOfObject: row];
+}
+
+- (NSGridColumn *) columnAtIndex: (NSInteger)index
+{
+  return [_columns objectAtIndex: index];
+}
+
+- (NSInteger) indexOfColumn: (NSGridColumn *)column
+{
+  return [_columns indexOfObject: column];
+}
+
+- (NSGridCell *) cellAtColumnIndex: (NSInteger)columnIndex rowIndex: (NSInteger)rowIndex
+{
+  NSUInteger idx = rowIndex * [self numberOfColumns] + columnIndex;
+  return [_cells objectAtIndex: idx];
+}
+
+- (NSGridCell *) cellForView: (NSView*)view
+{
+  NSGridCell *result = nil;
+
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      NSView *v = [c contentView];
+      if (v == view)
+        {
+          result = c;
+          break;
+        }
+    }
+  END_FOR_IN(_cells);
+
+  return result;
+}
+
+- (NSGridRow *) addRowWithViews: (NSArray *)views
+{
+  return [self insertRowAtIndex: [self numberOfRows]
+                      withViews: views];
+}
+
+- (NSGridRow *) _insertRowAtIndex: (NSInteger)index withCells: (NSArray *)cells
+{
+  NSGridRow *gr = [[NSGridRow alloc] init];
+
+  // Insert the row and release...
+  [_rows insertObject: gr atIndex: index];
+  [gr setGridView: self];
+  RELEASE(gr);
+
+  NSUInteger i = 0;
+  NSUInteger pos = index * [self numberOfColumns];
+  for(i = 0; i < [self numberOfColumns]; i++)
+    {
+      NSGridCell *c = [[NSGridCell alloc] init];
```

Comment:
You should not create a cell in this line.

Reply:
Removed